### PR TITLE
[CALCITE-5089] Allow GROUP BY ALL or DISTINCT set quantifier on GROUPING SETS

### PR DESCRIPTION
**Pull request by hannerwang**
This change will allow the following sql:
SELECT deptno, job, count() FROM emp GROUP BY ALL CUBE (deptno, job), ROLLUP (deptno, job)
and
SELECT deptno, job, count() FROM emp GROUP BY DISTINCT CUBE (deptno, job), ROLLUP (deptno, job)

The GROUP BY ALL is equivalent to GROUP BY, The GROUP BY DISTINCT will remove duplicates from cartesian product of CUBE and ROLLUP.

### CODE REVIEW

File: testkit/src/main/java/org/apache/calcite/sql/parser/SqlParserTest.java [hunk 1]
```diff
@@ -2122,6 +2122,29 @@ void checkPeriodPredicate(Checker checker) {
...
+    final String sql = "select deptno from emp\n"
+        + "group by all cube (a, b), rollup (a, b)";
+    final String expected = "SELECT `DEPTNO`\n"
+        + "FROM `EMP`\n"
+        + "GROUP BY CUBE(`A`, `B`), ROLLUP(`A`, `B`)";
+    sql(sql).ok(expected);
+
+    final String sql1 = "select deptno from emp\n"
+        + "group by distinct cube (a, b), rollup (a, b)";
+    final String expected1 = "SELECT `DEPTNO`\n"
```
Comment (chunweilei) [inline testkit/src/main/java/org/apache/calcite/sql/parser/SqlParserTest.java:2135]: Could you please add some negative case, such as `GROUP BY DISTINCT/ALL ()`?

Comment (hannerwang) [reply testkit/src/main/java/org/apache/calcite/sql/parser/SqlParserTest.java:2135]: Hi @chunweilei , I find the other tests have no this kind of edge cases, and do you mean `GROUP BY DISTINCT ()` ast should be unparsed as `GROUP BY ()`?

Comment (chunweilei) [reply testkit/src/main/java/org/apache/calcite/sql/parser/SqlParserTest.java:2135]: What's the behavior of PostgreSQL when meeting `GROUP BY DISTINCT ()`?

**Comment by chunweilei** [comment]
BTW, Julian left some comments in the JIRA issue. Could you please address them? Afterward, we can merge it.

File: testkit/src/main/java/org/apache/calcite/sql/parser/SqlParserTest.java (hunk 1 above)
Comment (hannerwang) [reply testkit/src/main/java/org/apache/calcite/sql/parser/SqlParserTest.java:2135]: I tried and that shows `GROUP BY DISTINCT ()` behaves same as `GROUP BY ()`.

**Comment by hannerwang** [comment]
> BTW, Julian left some comments in the JIRA issue. Could you please address them? Afterward, we can merge it.

Thanks @chunweilei , I submit some minor changes, please review.

**Review by chunweilei — APPROVED** [review]
LGTM.

File: core/src/main/codegen/templates/Parser.jj [hunk 2]
```diff
@@ -2509,8 +2510,7 @@ SqlNodeList GroupByOpt() :
     list = GroupingElementList() {
         if (distinct) {
             SqlNode groupByDistinct = SqlInternalOperators.GROUP_BY_DISTINCT.createCall(s.add(getPos()).pos(), list);
-            list = new ArrayList<SqlNode>();
-            list.add(groupByDistinct);
+            list = ImmutableNullableList.of(groupByDistinct);
```
Comment (julianhyde) [inline core/src/main/codegen/templates/Parser.jj:2513]: I think you can use ImmutableList, since groupByDistinct is never null
